### DISCUSSION TITLE
Pyodide in an Electron App

### DISCUSSION THREAD

**Discussion by byt3bl33d3r** [pyodide/pyodide#2248] [Q&A]
Heya,

I'm very much a Node/Electron n00b so apologies if this is a dumb question.

I'm currently trying to write an [Obsidian Plugin](https://obsidian.md/) (Obsidian is an Electron app) that embeds pyodide.

I have the following code:

```node
import { loadPyodide } from 'pyodide/pyodide.js';

export default class MyPlugin extends Plugin {
	settings: MyPluginSettings;

	async onload() {
		await this.loadSettings();

		// Pyodide
		const pyodide = await loadPyodide({
			indexURL : "https://cdn.jsdelivr.net/pyodide/v0.19.0/full/"
		});

		console.log( await pyodide.runPythonAsync("1 + 2") );

.... SNIP ....
```

Which throws the below error in the developer tools:
![image](https://user-images.githubusercontent.com/5151193/156494960-6926439b-541b-4323-a124-eae0adc26bce.png)

which is a lie cause that file does exist at that URL. Changing this URL to reference a local file does not work either.

Am i doing something obviously dumb here? Or is this some bug?

Thanks!

**Comment by hoodmane** [comment]
> Am i doing something obviously dumb here? Or is this some bug?

You aren't doing anything dumb. Pyodide feature detects electron as node, then it expects to be able to `import` the modules `"fs/promises"`, `"node-fetch"`, and `"vm"`. If it can't import those, then it requires patching. How does one usually load files in your environment?

**Reply by byt3bl33d3r** [reply]
TBH i'm not 100% sure as the documentation on Obsidian Plugins is a bit sparse or at least I haven't seen an official way of doing it. Looking at other plugins that read files it seems they're just using the standard `readFileSync()` from the fs module (example https://github.com/tillahoffmann/obsidian-jupyter/blob/master/main.ts#L73)

**Reply by hoodmane** [reply]
Well could you try `let fsmod = await import("fs");` and `import { readFile } from "fs/promises";`? If one of those doesn't work, that would give some extra info. It might be a problem with the dynamic es6 import.

**Reply by byt3bl33d3r** [reply]
Clarification: talking to a few people in the obsidian discord channel it seems like using the `vault.adapter` API is the way to read files as that would allow you to open files under the `.obsidian` folder which is where the plugin stuff would be stored. If it's outside a Vault (then you'd use fs APIs)

**Reply by byt3bl33d3r** [reply]
Ah ok so I added `let fsmod = await import("fs");` and now it complains about resolving `node-fetch`:
![image](https://user-images.githubusercontent.com/5151193/156500006-629dfe70-9edc-409e-92d1-d2ab7e727e75.png)

if i then install node fetch it just throws up:
![image](https://user-images.githubusercontent.com/5151193/156500080-9ba96956-070a-41d9-b860-5ee6df96afc4.png)

So i think it might be some weird import problem

**Reply by hoodmane** [reply]
Well you should be importing `node-fetch` through `node_modules/node-fetch/lib/index.es.js`, not through `node_modules/node-fetch/src/index.js`

**Reply by pjeby** [reply]
Obsidian is a *browser* application, not a node application.  On the desktop, Obsidian runs in an electron **renderer** process, which is to say a Chromium browser window.  (And it runs in a platform-specific browser on mobile devices.)  The desktop has a *limited* emulation of node APIs, but browser APIs are what should actually be used.  (And the node APIs are loaded via require(), not import, though build tools will generally convert imports to requires during the plugin build.)

The issue being encountered here is that pyodide's loader sees a valid `process` global (in the Electron environment), and then assumes it's running in node, even though the full node API is *not* available.  What's really needed is to be able to unconditionally use the browser API (e.g. plain `fetch` rather than node-fetch) and ignore node, since the absence of fs/promises is not the only difference between an Electron renderer and Node.  (And in any case, the mobile versions of Obsidian do not support even the minimal node API supplied by the desktop.)

Obsidian plugins are implemented as bundled .js files: that is, you use rollup or esbuild to roll everything your plugin does into a single CommonJS module that Obsidian loads at runtime, and which executes in a browser environment.  It is generally not recommended for plugins to access the filesystem directly, rather than through Obsidian's own APIs, since using node APIs makes the plugin unusable on mobile devices.

Depending on the builder used, it *might* be possible to fake out pyodide by telling the builder to inject an undefined `process` variable into the pyodide module.  But it would probably be better to just have the option to ask for the browser implementation with no autodetection.  Per the Zen of Python, "In the face of ambiguity, refuse the temptation to guess."  :wink:

**Reply by hoodmane** [reply]
Well an approach that would work with the existing version of Pyodide is to set `process.browser = "Obsidian"` or something.
```js
const IN_NODE = 
  typeof process !== "undefined" &&
  process.release &&
  process.release.name === "node" &&
  typeof process.browser ===
    "undefined"; 
```
We could potentially add a `loadPyodide` config option that can set the runtime (current options are browser main thread, browser classic worker, browser module worker, or node).

**Reply by hoodmane** [reply]
We could also add to the `IN_NODE` test: `&& !!process.versions.electron`.

**Reply by pjeby** [reply]
My understanding is that process.browser is deprecated and that `window` (or `self`) existing is recommended for detecting a browser environment.  Detecting the lack of electron is of course also an option, but I don't know it would be the *correct* option in all cases, since Electron also has the non-browser main process, and in any case Electron apps may vary in what they're trying to *do*.

For example, an Electron app that uses pyodide in its *main* process might well want to bundle the wasm code inside its .asar archive, for which the `fs` module (specifically, the Electron-patched version) would then definitely be required!

So I would reiterate the suggestion to allow explicitly selecting the runtime (or providing a custom implementation thereof) as it eliminates the need for guessing, and then needing to add even *more* guessing rules when somebody else comes along with another new use case or platform.

**Comment by Parmarjaimin568** [comment]
when Pyodid loaded in my Python.JSX file when error Failed to initialize Pyodide: Failed to resolve module specifier "fs/promises". Relative references must start with either "/", "./", or "../".
